### DISCUSSION TITLE
Support passphrase protected ssh keys

### DISCUSSION THREAD

**Discussion by rakuten-JO** [cargo-generate/cargo-generate#1184] [Q&A] [answered]
**Describe the bug**
I want to start a project:
```shell
cargo generate --git https://github.com/rustwasm/wasm-pack-template.git
```
And I get:
```shell
Error: Please check if the Git user / repository exists.

Caused by:
    Failed to retrieve list of SSH authentication methods: Failed getting response; class=Ssh (23); code=Auth (-16)
```

**Desktop (please complete the following information):**
 - OS: [MacOS]
 - Version: [Mac Os Big Sur version 11.4]
 - `rustc --version`: rustc 1.65.0 (897e37553 2022-11-02)
 - `cargo generate --version`: cargo generate-generate 0.17.3

**Additional context**
it is fine in my Windows PC, but i guess it may be the problem of my .git config.
```shell
# Personal account, - the default config
Host github.com-personal
   HostName github.com
   User git
   IdentityFile ~/.ssh/id_rsa
   
# Work account
Host github.com-organization
   HostName github.com
   User git
   IdentityFile ~/.ssh/id_ed25519
```
And I want to know how to fix it?

**Comment by sassman** [comment]
Interesting, are your keys protected by a passphrase?

As it seems it is using git over ssh even if you've given it a http url.

When you comment out the .gitconfig lines you've mentioned then I guess it does work, right?

**Comment by rakuten-JO** [comment]
> Interesting, are your keys protected by a passphrase?
> 
> As it seems it is using git over ssh even if you've given it a http url.
> 
> When you comment out the .gitconfig lines you've mentioned then I guess it does work, right?

Yes, in my Windows PC, I do not have the config file i mentioned above, and it worked, but I do not know the reason of it.
My git config is similar with this ->`Multiple Github Accounts on a single Machine`:
https://gist.github.com/rahularity/86da20fe3858e6b311de068201d279e3

**Comment by sassman** [comment]
@rakuten-JO can you please try the approach from the comment below and see if that would solve your issue?

https://github.com/rust-lang/cargo/issues/3381#issuecomment-725980194

**Comment by moridinga** [comment]
> @rakuten-JO can you please try the approach from the comment below and see if that would solve your issue?
> 
> [rust-lang/cargo#3381 (comment)](https://github.com/rust-lang/cargo/issues/3381#issuecomment-725980194)

I've tried that approach (`git-fetch-with-cli = true` in the config.toml for cargo), and it doesn't work for cargo generate. I would have expected it to, but it does not. That would be my feature request, get cargo generate to use/respect this option do do git ops via the cli instead of using libgit2.

**Comment by stappersg** [comment]
Augustus 2023, it works for me ...
```text
stappers@paddy:/tmp
$ cargo generate --git https://github.com/rustwasm/wasm-pack-template.git
🤷   Project Name: foo
🔧   Destination: /tmp/foo ...
🔧   project-name: foo ...
🔧   Generating template ...
🔧   Moving generated files into: `/tmp/foo`...
Initializing a fresh Git repository
✨   Done! New project created /tmp/foo
stappers@paddy:/tmp
$ ls /tmp/foo
Cargo.toml  LICENSE_APACHE  LICENSE_MIT  README.md  src  tests
stappers@paddy:/tmp
$ cargo generate --version
cargo generate-generate 0.18.3
stappers@paddy:/tmp
$ 
```

@moridinga  please retry with latest `cargo generate` and update this issue.
@rakuten-JO please retry with latest `cargo generate`, update issue and maybe close it.

**Comment by sassman** [comment, marked as answer]
Adding the ssh key to ssh-agent should in general work

```shell
eval $(ssh-agent)
ssh-add <path-to-your-private-key>
cargo ...
```

or on mac

```shell
eval $(ssh-agent)
ssh-add --apple-use-keychain <path-to-your-private-key>
cargo ...
```

**Reply by sassman** [reply]
see also how to setup ssh-agent on windows: https://github.com/cargo-generate/cargo-generate/discussions/653#discussion-4148438
